Add Dropbox backup integration

bdr99 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new integration for Dropbox, integrating with the backup feature of Home Assistant

Note: As discussed with @joostlek, I will be coordinating with the Open Home Foundation to set up Cloud Account Linking for this integration (subject to Dropbox's approval of the application) before it is released in a stable HA release.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/41560
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

bdr99 committed:
Add Dropbox integration (WIP)

bdr99 committed:
Merge branch 'dev' into dropbox-integration

bdr99 committed:
Merge branch 'dev' into dropbox-integration

bdr99 committed:
add tests

bdr99 committed:
small fixes

bdr99 committed:
small fixes

cajomferro commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,225 @@
+            return
+
+        if not metadata.is_folder:
+            raise HomeAssistantError("Home Assistant exists as a file, not a folder")
```
Kind of agree with this comment. The current error message might be misleading.

cajomferro commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,225 @@
+                                json.loads(property_group.fields[0].value)
+                            )
+                        )
+                        break  # Found matching template, no need to check other groups
```
consider moving this code to a helper method (see my other comment below)

cajomferro reviewed (commented):
The integration acts as a backup agent, allowing Home Assistant to use a Dropbox account as a storage location for backups. For authentication, it uses OAuth2 with PKCE for secure, user-authorized access to the Dropbox API, leveraging Home Assistant's application credentials system for storing the client ID and secret. Overall, the integration and the code itself are well structured and follows HASS standards.

The integration is currently at the Bronze quality level but maybe it could achieve the Silver level without too much effort. Significantly test coverage has been added and there are other params that your integration already achieves as well so remember to evaluate all rules, as mentioned by @joostlek. Another possible improvement is logging for unavailable states: the API calls in api.py and backup.py should include logging to record errors when Dropbox is unreachable or authentication fails, which helps with troubleshooting.

bdr99 committed:
Address review comments

bdr99 commented on homeassistant/components/dropbox/quality_scale.yaml:
```diff
@@ -0,0 +1,74 @@
+  log-when-unavailable: todo
+  parallel-updates: todo
+  reauthentication-flow: todo
+  test-coverage: todo
```
The integration actually doesn't currently reach the 95% coverage threshold that is required by this rule.

bdr99 commented on homeassistant/components/dropbox/__init__.py:
```diff
@@ -0,0 +1,52 @@
+
+    def async_notify_backup_listeners() -> None:
+        for listener in hass.data.get(DATA_BACKUP_AGENT_LISTENERS, []):
+            listener()
```
I checked other integrations and don't see any of them doing this. I would prefer to follow the example of other integrations that are already using this pattern.

bdr99 commented on homeassistant/components/dropbox/quality_scale.yaml:
```diff
@@ -0,0 +1,74 @@
+  # Platinum
+  async-dependency: todo
+  inject-websession: todo
+  strict-typing: todo
```
Updated.

bdr99 commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,225 @@
+                        )
+                        if backup.backup_id == backup_id:
+                            await self._api.delete_file(f"/{FOLDER_NAME}/{file.name}")
+                            return
```
Fixed.

zweckj commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,209 @@
+)
+
+
+class DropboxClient:
```
I don't like the mixing of responsibilities in one file. Move the auth things to `auth.py`

bdr99 replied:
Fixed.

zweckj commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,209 @@
+    async def _async_get_backups(self) -> list[tuple[AgentBackup, str]]:
+        """Get backups and their corresponding file names."""
+        property_template_id = await self.async_get_property_template_id()
+
```
why do we need to ensure the folder exists during `list`, but not during `get`?

zweckj commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,209 @@
+
+        file_stream = await open_stream()
+
+        await self._api.upload_file(
```
those are in general not error handled, is there a reason for that?

zweckj commented on homeassistant/components/dropbox/config_flow.py:
```diff
@@ -0,0 +1,58 @@
+        auth = AsyncConfigFlowAuth(async_get_clientsession(self.hass), access_token)
+
+        client = DropboxClient(auth)
+        account_info = await client.async_get_account_info()
```
I don't know how permission granting works in Dropbox, but could we come back from the OAuth flow with a token that does not have permissions to create the required files/folders, e.g. by the user only granting a part of the requested permissions?

bdr99 replied:
I just checked and the auth flow doesn’t provide an option for the user to only grant a partial set of permissions. They have to grant all permissions in order to finish the flow.

MartinHjelmare commented on homeassistant/components/dropbox/manifest.json:
```diff
@@ -0,0 +1,12 @@
+  "codeowners": ["@bdr99"],
+  "config_flow": true,
+  "dependencies": ["application_credentials"],
+  "documentation": "https://www.home-assistant.io/integrations/dropbox",
```
Regarding this suggestion:

Use a comma to separate lines (objects).

And take care of whitespace. Looks like there's one preceding space too many on line 8 in the suggested change.

bdr99 replied:
Fixed.

bdr99 committed:
Address review comments

bdr99 commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,209 @@
+    def __init__(
+        self,
+        websession: ClientSession,
+        oauth_session: config_entry_oauth2_flow.OAuth2Session,
```
Fixed.

bdr99 commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,209 @@
+    async def _async_ensure_folder_exists(self) -> None:
+        """Ensure the folder exists."""
+        try:
+            metadata = await self._api.get_metadata(f"/{FOLDER_NAME}")
```
I had the same thought, but unfortunately it seems that App Folder scoped apps are not allowed to work with property templates. I get this error: `Error in call to API function "file_properties/templates/list_for_user": Your API app is an "App Folder" app. It is not allowed to access this API function.`

zweckj replied:
since we cannot use them (see above) let's re-explore that please

bdr99 replied:
Updated to use an App Folder app scope. With an App Folder app scope, all file access via the api is relative to the app folder instead of the dropbox root, so I removed the functionality for ensuring the folder exists and updated the path to just `""` instead of `"/Home Assistant"`. I also updated the docs PR accordingly.

bdr99 commented on homeassistant/components/dropbox/quality_scale.yaml:
```diff
@@ -0,0 +1,116 @@
+  docs-configuration-parameters:
+    status: exempt
+    comment: Integration does not have any configuration parameters.
+  docs-installation-parameters:
```
Marked as done and added instructions to the docs PR explaining how to create and use your own application credentials.

zweckj reviewed (changes requested):
setting this back until we have the comments solved

bdr99 committed:
Merge branch 'home-assistant:dev' into dropbox-integration

bdr99 committed:
Add a test for removing the backup agents listener

bdr99 committed:
Use metadata files instead of property groups

bdr99 committed:
Add decorator to handle common backup errors

bdr99 committed:
Handle unknown error (internet outage) during setup

bdr99 committed:
Improve test coverage

bdr99 committed:
Switch to app folders instead of full dropbox access

bdr99 commented on homeassistant/components/dropbox/__init__.py:
```diff
@@ -0,0 +1,52 @@
+async def async_setup_entry(hass: HomeAssistant, entry: DropboxConfigEntry) -> bool:
+    """Set up Dropbox from a config entry."""
+    oauth2_implementation = await async_get_config_entry_implementation(hass, entry)
+    oauth2_session = OAuth2Session(hass, entry, oauth2_implementation)
```
Fixed. I tested this by disconnecting the internet and verifying the new error was correctly thrown during setup.

bdr99 commented on homeassistant/components/dropbox/backup.py:
```diff
@@ -0,0 +1,151 @@
+            DropboxFileOrFolderNotFoundException,
+            DropboxUnknownException,
+        ) as err:
+            raise BackupAgentError("Failed to download backup") from err
```
Updated.

bdr99 committed:
python-dropbox-api 0.1.3

bdr99 commented:
@zweckj Thanks for the great suggestions, the code is a lot simpler now with the migration to app folder scope and removal of property fields. I addressed all your comments.

bdr99 committed:
raise BackupNotFound on DropboxFileOrFolderNotFoundException

bdr99 committed:
Start reauth on auth error in backup agent

bdr99 committed:
Remove integration name from log

bdr99 committed:
Update quality scale

zweckj commented on homeassistant/components/dropbox/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+            return self.async_show_form(step_id="reauth_confirm")
+        return await self.async_step_user()
+
+    async def async_oauth_create_entry(self, data: dict[str, Any]) -> ConfigFlowResult:
```
move that above the reauth things for better ordering

zweckj commented on homeassistant/components/dropbox/api.py:
```diff
@@ -0,0 +1,117 @@
+    yield content.encode()
+
+
+class DropboxClient:
```
TBH I would just move the code here inline into the backup.py now. It's all so very tightly coupled together, that I don't see a reason why we should add an extra interface layer.

zweckj commented on tests/components/dropbox/test_api.py:
this is another reason why I don't like the separate api. We shouldn't need to test that separately, this should be all covered  by the backup tests

bdr99 commented on homeassistant/components/dropbox/quality_scale.yaml:
```diff
@@ -0,0 +1,114 @@
+    comment: Integration does not have any entities.
+  reconfiguration-flow:
+    status: exempt
+    comment: Integration does not have any configuration options.
```
What is there to reconfigure for this integration? According to the [docs](https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#reconfigure), reconfiguration flows should not be used for handling auth issues, but the config flow just consists of an auth flow and nothing else.

bdr99 replied:
Sorry I'm still confused. Other than auth (since the docs say that reconfiguration flows should not be used for auth issues), what configuration data is there to change?

bdr99 committed:
reorder methods in OAuth2FlowHandler

bdr99 committed:
handle ImplementationUnavailableError

bdr99 committed:
Add mock_setup_entry fixture

bdr99 committed:
Move _async_string_iterator outside try

bdr99 commented on homeassistant/components/dropbox/__init__.py:
```diff
@@ -0,0 +1,54 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: DropboxConfigEntry) -> bool:
+    """Set up Dropbox from a config entry."""
+    oauth2_implementation = await async_get_config_entry_implementation(hass, entry)
```
Fixed.

bdr99 committed:
Remove api.py layer, move logic into backup.py

bdr99 committed:
Remove test_auth.py and cover auth through init and config flow tests

bdr99 committed:
Mark reconfiguration-flow as todo

bdr99 committed:
address review comments

bdr99 commented on tests/components/dropbox/test_config_flow.py:
```diff
@@ -0,0 +1,205 @@
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == ACCOUNT_EMAIL
+    assert result["data"]["token"]["access_token"] == "mock-access-token"
+    assert result["context"]["unique_id"] == ACCOUNT_ID
```
Updated.

zweckj reviewed (changes requested):
last things from my end

bdr99 commented on homeassistant/components/dropbox/backup.py:
```diff
@@ -0,0 +1,221 @@
+            _LOGGER.debug("Full error: %s", err, exc_info=True)
+            raise BackupAgentError(
+                f"Failed to {func.__name__.removeprefix('async_').replace('_', ' ')}"
+            ) from err
```
The library would catch a `TimeoutError` and convert it to a `DropboxUnknownException`

bdr99 committed:
rename classes to be specific to dropbox

bdr99 committed:
rename config_entry fixture to mock_config_entry

bdr99 commented on tests/components/dropbox/conftest.py:
```diff
@@ -0,0 +1,114 @@
+
+
+@pytest.fixture
+def config_entry() -> MockConfigEntry:
```
Updated.

zweckj commented on homeassistant/components/dropbox/backup.py:
```diff
@@ -0,0 +1,221 @@
+
+            metadata_stream = self._api.download_file(f"/{metadata_file.name}")
+            raw = b"".join([chunk async for chunk in metadata_stream])
+            backup = AgentBackup.from_dict(json.loads(raw))
```
this sounds like valid feedback

zweckj commented on homeassistant/components/dropbox/strings.json:
```diff
@@ -0,0 +1,35 @@
+      "oauth_unauthorized": "[%key:common::config_flow::abort::oauth2_unauthorized%]",
+      "reauth_successful": "[%key:common::config_flow::abort::reauth_successful%]",
+      "user_rejected_authorize": "[%key:common::config_flow::abort::oauth2_user_rejected_authorize%]",
+      "wrong_account": "Wrong account: Please authenticate with the right account."
```
I like that one

zweckj commented on homeassistant/components/dropbox/strings.json:
```diff
@@ -0,0 +1,35 @@
+        "title": "[%key:common::config_flow::title::oauth2_pick_implementation%]"
+      },
+      "reauth_confirm": {
+        "description": "The Dropbox integration needs to re-authenticate your account",
```
also fair

bdr99 committed:
address review comments
